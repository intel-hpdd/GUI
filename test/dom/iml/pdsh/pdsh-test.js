// @flow

import { render } from "inferno";

const waitForDebounce = () => {
  return new Promise(res => {
    setTimeout(() => {
      res(true);
    }, 200);
  });
};

describe("PDSH component", () => {
  let el: HTMLElement,
    pdshPlaceholder: string,
    body: HTMLBodyElement,
    pdshInput: HTMLInputElement,
    hostnamePopoverIcon: HTMLElement,
    Pdsh;

  beforeEach(() => {
    pdshPlaceholder = "placeholder text";

    jest.mock("../../../../source/iml/environment.js", () => ({
      HELP_TEXT: {
        pdsh_placeholder: "pdsh placeholder text"
      }
    }));

    Pdsh = require("../../../../source/iml/pdsh/pdsh.js").default;

    body = (document.body: any);
    el = document.createElement("div");
    body.appendChild(el);

    render(<Pdsh pdshPlaceholder={pdshPlaceholder} />, el);
  });

  describe("general operation", () => {
    let inputEvent: KeyboardEvent;
    beforeEach(() => {
      inputEvent = new KeyboardEvent("input");
      pdshInput = (el.querySelector('input[name="pdsh"]'): any);
      hostnamePopoverIcon = (el.querySelector(".hostname-popover"): any);
    });

    describe("without input", () => {
      it("should render", () => {
        expect(el).toMatchSnapshot();
      });
    });

    describe("with input", () => {
      it("should render", async () => {
        pdshInput.value = "oss[1,2].local";
        pdshInput.dispatchEvent(inputEvent);

        await waitForDebounce();
        expect(el).toMatchSnapshot();
      });
    });

    describe("with input error", () => {
      it("should render", async () => {
        pdshInput.value = "oss[1,].local";
        pdshInput.dispatchEvent(inputEvent);

        await waitForDebounce();
        expect(el).toMatchSnapshot();
      });
    });

    describe("with initial value", () => {
      it("should render", () => {
        render(null, el);
        render(<Pdsh pdshPlaceholder={pdshPlaceholder} pdshInitial={"mds[1,2].local"} />, el);
        hostnamePopoverIcon = (el.querySelector(".hostname-popover"): any);
        hostnamePopoverIcon.click();

        expect(el).toMatchSnapshot();
      });
    });

    describe("clicking the hostname popover", () => {
      it("should render", async () => {
        pdshInput.value = "oss[1,2].local";
        pdshInput.dispatchEvent(inputEvent);
        hostnamePopoverIcon.click();

        await waitForDebounce();
        expect(el).toMatchSnapshot();
      });
    });

    describe("with debounce", () => {
      beforeEach(() => {
        pdshInput.value = "o";
        pdshInput.dispatchEvent(inputEvent);
        pdshInput.value = "os";
        pdshInput.dispatchEvent(inputEvent);
        pdshInput.value = "oss";
        pdshInput.dispatchEvent(inputEvent);
        pdshInput.value = "oss[";
        pdshInput.dispatchEvent(inputEvent);
        pdshInput.value = "oss[1";
        pdshInput.dispatchEvent(inputEvent);
        pdshInput.value = "oss[1,],";
        pdshInput.dispatchEvent(inputEvent);
        pdshInput.value = "oss[1,2";
        pdshInput.dispatchEvent(inputEvent);
        pdshInput.value = "oss[1,2]";
        pdshInput.dispatchEvent(inputEvent);
      });

      describe("before debounce fires", () => {
        it("should render with non-debounced data", () => {
          hostnamePopoverIcon.click();
          expect(el).toMatchSnapshot();
        });
      });

      describe("after debounce", () => {
        it("should render with debounced data after the timeout", async () => {
          hostnamePopoverIcon.click();

          await waitForDebounce();
          expect(el).toMatchSnapshot();
        });
      });
    });
  });
});
