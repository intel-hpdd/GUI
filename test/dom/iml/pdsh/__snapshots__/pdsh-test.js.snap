// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDSH component general operation clicking the hostname popover should render 1`] = `
<div>
  <div
    class="input-group"
  >
    <span
      style="position: relative;"
    >
      <div
        class="input-group-addon tooltip-container tooltip-hover hostname-popover"
        popoverbutton="true"
      >
        <i
          class="fa activate-popover fa-list-ul"
        >
          <div
            class="tooltip inferno-tt right medium "
          >
            <div
              class="tooltip-arrow"
            />
            <div
              class="tooltip-inner"
            >
              <span>
                Click for expanded hostlist expression.
              </span>
            </div>
          </div>
        </i>
      </div>
      <div
        class="fade popover bottom in"
      >
        <div
          class="arrow"
        />
        <h3
          class="popover-title"
        >
          Hosts
        </h3>
        <div
          class="popover-content"
        >
          <ul
            class="well"
          >
            <li>
              <span>
                oss1..2.local
              </span>
            </li>
          </ul>
        </div>
      </div>
    </span>
  </div>
  <input
    class="form-control"
    name="pdsh"
    placeholder="placeholder text"
    required=""
    type="search"
  />
  
</div>
`;

exports[`PDSH component general operation with debounce after debounce should render with debounced data after the timeout 1`] = `
<div>
  <div
    class="input-group"
  >
    <span
      style="position: relative;"
    >
      <div
        class="input-group-addon tooltip-container tooltip-hover hostname-popover"
        popoverbutton="true"
      >
        <i
          class="fa activate-popover fa-list-ul"
        >
          <div
            class="tooltip inferno-tt right medium "
          >
            <div
              class="tooltip-arrow"
            />
            <div
              class="tooltip-inner"
            >
              <span>
                Click for expanded hostlist expression.
              </span>
            </div>
          </div>
        </i>
      </div>
      <div
        class="fade popover bottom in"
      >
        <div
          class="arrow"
        />
        <h3
          class="popover-title"
        >
          Hosts
        </h3>
        <div
          class="popover-content"
        >
          <ul
            class="well"
          >
            <li>
              <span>
                oss1..2
              </span>
            </li>
          </ul>
        </div>
      </div>
    </span>
  </div>
  <input
    class="form-control"
    name="pdsh"
    placeholder="placeholder text"
    required=""
    type="search"
  />
  
</div>
`;

exports[`PDSH component general operation with debounce before debounce fires should render with non-debounced data 1`] = `
<div>
  <div
    class="input-group"
  >
    <span
      style="position: relative;"
    >
      <div
        class="input-group-addon tooltip-container tooltip-hover hostname-popover"
        popoverbutton="true"
      >
        <i
          class="fa activate-popover fa-list-ul"
        >
          <div
            class="tooltip inferno-tt right medium "
          >
            <div
              class="tooltip-arrow"
            />
            <div
              class="tooltip-inner"
            >
              <span>
                Click for expanded hostlist expression.
              </span>
            </div>
          </div>
        </i>
      </div>
      <div />
    </span>
  </div>
  <input
    class="form-control"
    name="pdsh"
    placeholder="placeholder text"
    required=""
    type="search"
  />
  <div
    class="tooltip inferno-tt bottom medium error-tooltip"
  >
    <div
      class="tooltip-arrow"
    />
    <div
      class="tooltip-inner"
    >
      <ul>
        <li>
          Expression required.
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`PDSH component general operation with initial value should render 1`] = `
<div>
  <div
    class="input-group"
  >
    <span
      style="position: relative;"
    >
      <div
        class="input-group-addon tooltip-container tooltip-hover hostname-popover"
        popoverbutton="true"
      >
        <i
          class="fa activate-popover fa-list-ul"
        >
          <div
            class="tooltip inferno-tt right medium "
          >
            <div
              class="tooltip-arrow"
            />
            <div
              class="tooltip-inner"
            >
              <span>
                Click for expanded hostlist expression.
              </span>
            </div>
          </div>
        </i>
      </div>
      <div
        class="fade popover bottom in"
      >
        <div
          class="arrow"
        />
        <h3
          class="popover-title"
        >
          Hosts
        </h3>
        <div
          class="popover-content"
        >
          <ul
            class="well"
          >
            <li>
              <span>
                mds1..2.local
              </span>
            </li>
          </ul>
        </div>
      </div>
    </span>
  </div>
  <input
    class="form-control"
    name="pdsh"
    placeholder="placeholder text"
    required=""
    type="search"
  />
  
</div>
`;

exports[`PDSH component general operation with input error should render 1`] = `
<div>
  <div
    class="input-group"
  >
    <span
      style="position: relative;"
    >
      <div
        class="input-group-addon tooltip-container tooltip-hover hostname-popover"
        popoverbutton="true"
      >
        <i
          class="fa activate-popover fa-list-ul"
        >
          <div
            class="tooltip inferno-tt right medium "
          >
            <div
              class="tooltip-arrow"
            />
            <div
              class="tooltip-inner"
            >
              <span>
                Click for expanded hostlist expression.
              </span>
            </div>
          </div>
        </i>
      </div>
      <div />
    </span>
  </div>
  <input
    class="form-control"
    name="pdsh"
    placeholder="placeholder text"
    required=""
    type="search"
  />
  <div
    class="tooltip inferno-tt bottom medium error-tooltip"
  >
    <div
      class="tooltip-arrow"
    />
    <div
      class="tooltip-inner"
    >
      <ul>
        <li>
          Range is not in the proper format.
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`PDSH component general operation with input should render 1`] = `
<div>
  <div
    class="input-group"
  >
    <span
      style="position: relative;"
    >
      <div
        class="input-group-addon tooltip-container tooltip-hover hostname-popover"
        popoverbutton="true"
      >
        <i
          class="fa activate-popover fa-list-ul"
        >
          <div
            class="tooltip inferno-tt right medium "
          >
            <div
              class="tooltip-arrow"
            />
            <div
              class="tooltip-inner"
            >
              <span>
                Click for expanded hostlist expression.
              </span>
            </div>
          </div>
        </i>
      </div>
      
    </span>
  </div>
  <input
    class="form-control"
    name="pdsh"
    placeholder="placeholder text"
    required=""
    type="search"
  />
  
</div>
`;

exports[`PDSH component general operation without input should render 1`] = `
<div>
  <div
    class="input-group"
  >
    <span
      style="position: relative;"
    >
      <div
        class="input-group-addon tooltip-container tooltip-hover hostname-popover"
        popoverbutton="true"
      >
        <i
          class="fa activate-popover fa-list-ul"
        >
          <div
            class="tooltip inferno-tt right medium "
          >
            <div
              class="tooltip-arrow"
            />
            <div
              class="tooltip-inner"
            >
              <span>
                Click for expanded hostlist expression.
              </span>
            </div>
          </div>
        </i>
      </div>
      <div />
    </span>
  </div>
  <input
    class="form-control"
    name="pdsh"
    placeholder="placeholder text"
    required=""
    type="search"
  />
  <div
    class="tooltip inferno-tt bottom medium error-tooltip"
  >
    <div
      class="tooltip-arrow"
    />
    <div
      class="tooltip-inner"
    >
      <ul>
        <li>
          Expression required.
        </li>
      </ul>
    </div>
  </div>
</div>
`;
